SQL: IN operator incorrectly handling specified date format (#58932)

wuyinan0126 opened the pull request:
Related to #58932

In `IN` operator, the ExpressionTranslator<In> translates expression into terms query, but when the expression is date format, the terms query get unexpected results. This commit changes the terms query into bool should query when data type is `DATETIME`.

The original translation of `IN ('2000-11-11 11:11:11'::timestamp)` was below, which got the error `failed to parse date field [2000-11-11T11:11:11Z] with format [yyyy-MM-dd' 'HH:mm:ss]: [failed to parse date field [2000-11-11T11:11:11Z] with format [yyyy-MM-dd' 'HH:mm:ss]]`, beacause the terms format is not the same as defined:

```
{
  "query" : {
    "terms" : {
      "datetime0" : [
        "2000-11-11T11:11:11.000Z"
      ],
      "boost" : 1.0
    }
  }
}
```

In this commit, the `IN ('2000-11-11 11:11:11'::timestamp)` translates to:
```
{
  "query" : {
    "range" : {
      "datetime0" : {
        "from" : "2000-11-11T11:11:11.000Z",
        "to" : "2000-11-11T11:11:11.000Z",
        "include_lower" : true,
        "include_upper" : true,
        "time_zone" : "Z",
        "format" : "strict_date_time",
        "boost" : 1.0
      }
    }
  }
}
```

And the `IN ('2000-11-11 11:11:11'::timestamp, '2000-11-11 22:22:22'::timestamp)` translate to:
```
{
 "query" : {
    "bool" : {
      "should" : [
        {
          "range" : {
            "datetime0" : {
              "from" : "2000-11-11T11:11:11.000Z",
              "to" : "2000-11-11T11:11:11.000Z",
              "include_lower" : true,
              "include_upper" : true,
              "time_zone" : "Z",
              "format" : "strict_date_time",
              "boost" : 1.0
            }
          }
        },
        {
          "range" : {
            "datetime0" : {
              "from" : "2000-11-11T22:22:22.000Z",
              "to" : "2000-11-11T22:22:22.000Z",
              "include_lower" : true,
              "include_upper" : true,
              "time_zone" : "Z",
              "format" : "strict_date_time",
              "boost" : 1.0
            }
          }
        }
      ],
      "boost" : 1.0
    }
  }
}
```

Finally, the problem is solved:

```
SELECT datetime0 FROM idx WHERE datetime0 IN ('2000-11-11 11:11:11'::timestamp)

       datetime0        
------------------------
2000-11-11T11:11:11.000Z
```

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

In `IN` operator, the ExpressionTranslator<In> translates expression into terms query, but when the expression is date format, the terms query get unexpected results. This commit changes the terms query into bool should query when data type is `DATETIME`.

wuyinan0126 commented:
@elasticmachine Please help review this commit~

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

wuyinan0126 commented:
@bpintea got it~

matriv commented:
@wuyinan0126 Thank you for picking this up!

The solution overall is correct, I'd like to ask you though if you could somehow reuse the logic for literal binary comparisons (simple `=`, `>`, ... etc.) here: https://github.com/elastic/elasticsearch/blob/2b94f88639dbf142576a1f24af34156b033190b2/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java#L238 which also take care of time only literals. Moreover, I'd like to ask you to add some unit test in `QueryTranslatorTests` which shows the new translation with the RangeQueries joined with BoolQueries, and maybe some integration tests at the bottom of `filter.sql-spec` where the IN operator is tested.

bpintea commented:
@wuyinan0126, would you be interested and have the time to finish this PR with the above feedback?

wuyinan0126 commented:
@bpintea sorry to see it so late, and I will work on it soon.

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Reuse the logic for literal binary comparisons and add the unitTests and integTests

Commit:
SQL: IN operator incorrectly handling specified date format (#58932)

Add the unitTests and integTests on `col::time in ('...'::time)`

wuyinan0126 commented:
@matriv Thanks for your advise, and I have tested and add time only literals case.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:
```diff
@@ -358,7 +360,23 @@ public static Query doTranslate(In in, TranslatorHandler handler) {
 
-                q = new TermsQuery(in.source(), fa.exactAttribute().name(), set);
+                Object value = valueOf(list.get(0));
+                if (value instanceof ZonedDateTime || value instanceof OffsetTime) {
+                    DateFormatter formatter;
+                    q = null;
+                    if (value instanceof ZonedDateTime) {
+                        formatter = DateFormatter.forPattern(DATE_FORMAT);
+                    } else {
+                        formatter = DateFormatter.forPattern(TIME_FORMAT);
```
This is not correct, since the branch will never be accessed with an `OffsetTime` instance: as you can see in `testTranslateInExpression_WhereClause_TimeValue` in [your own commit](https://github.com/elastic/elasticsearch/pull/60377/files/43dd7e29fc555a1e5da25d1d27243ed357485bb1..f48691a17c76a0672573866bf1ac9c44c3972fb7), the field's `::time` casting is translated as a script, so the flow will not visit this branch in that case.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/resources/filter.sql-spec:
```diff
@@ -129,3 +129,11 @@ whereWithInAndNullHandling1
+SELECT birth_date FROM "test_emp" WHERE birth_date IN ('1959-12-25T12:12:12.000Z'::datetime, '1959-07-23T12:12:12.000Z'::datetime);
+
+whereWithInAndTimeValue
+SELECT birth_date FROM "test_emp" WHERE birth_date::time IN ('12:12:12.000'::time);
```
This new addition is good, but if you actually the run the tests, it will fail due to #63521, which will need fixing before such a test can be merged.

bpintea commented:
Thanks for your new commits. As you might have seen on this PR's references, I've opened a new #63483 PR based on this one, since we weren't sure if you were still interested in continuing the work.
As you can see, you are mentioned as co-authoring the new PR and your commits have been kept, so proper credit is provided. So I hope you won't mind if we close this PR and continue the work on the new one.
Thank you again for your contribution!

bpintea commented:
Addressed in #63483.
